src/blocking.rs

```diff
@@ -0,0 +1,2830 @@
+            .ok_or_else(|| {
+                OciDistributionError::GenericError(Some(
+                    "could not clone request builder".to_string(),
+                ))
+            })?
+            .header("Accept", Vec::from(accept).join(", "));
+
+        Ok(RequestBuilderWrapper {
+            client: self.client,
+            request_builder,
+        })
+    }
+
+    /// Updates request as necessary for authentication.
+    ///
+    /// If the struct has Some(bearer), this will insert the bearer token in an
+    /// Authorization header. It will also set the Accept header, which must
+    /// be set on all OCI Registry requests. If the struct has HTTP Basic Auth
+    /// credentials, these will be configured.
+    fn apply_auth(
+        &mut self,
+        image: &Reference,
+        op: RegistryOperation,
+    ) -> Result<RequestBuilderWrapper> {
+        let mut headers = HeaderMap::new();
+
+        if let Some(token) = self.client.get_auth_token(image, op) {
+            match token {
+                RegistryTokenType::Bearer(token) => {
+                    debug!("Using bearer token authentication.");
+                    headers.insert("Authorization", token.bearer_token().parse().unwrap());
+                }
+                RegistryTokenType::Basic(username, password) => {
+                    debug!("Using HTTP basic authentication.");
+                    return Ok(RequestBuilderWrapper {
+                        client: self.client,
+                        request_builder: self
+                            .request_builder
+                            .try_clone()
+                            .ok_or_else(|| {
+                                OciDistributionError::GenericError(Some(
+                                    "could not clone request builder".to_string(),
+                                ))
+                            })?
+                            .headers(headers)
+                            .basic_auth(username.to_string(), Some(password.to_string())),
+                    });
+                }
+            }
+        }
+        Ok(RequestBuilderWrapper {
+            client: self.client,
+            request_builder: self
+                .request_builder
+                .try_clone()
+                .ok_or_else(|| {
+                    OciDistributionError::GenericError(Some(
+                        "could not clone request builder".to_string(),
+                    ))
+                })?
+                .headers(headers),
+        })
+    }
+}
+
+/// The encoding of the certificate
+#[derive(Debug, Clone)]
+pub enum CertificateEncoding {
+    #[allow(missing_docs)]
+    Der,
+    #[allow(missing_docs)]
+    Pem,
+}
+
+/// A x509 certificate
+#[derive(Debug, Clone)]
+pub struct Certificate {
+    /// Which encoding is used by the certificate
+    pub encoding: CertificateEncoding,
+
+    /// Actual certificate
+    pub data: Vec<u8>,
+}
+
+/// A client configuration
+pub struct ClientConfig {
```

Comment:
We should do whatever we need to here to make sure this is shared across both blocking and non-blocking clients. Pretty much everything here until the tests can be shared across clients